### PR TITLE
ceph: Use aggregated ClusterRoles

### LINKED ISSUE #2634: Rook should use aggregated roles to improve upgrades
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
During the upgrade between releases of Rook, we have to remove and create new Roles/ClusterRoles whenever there is an RBAC setting that changes, whether rook needs a new privilege or a privilege changes to a new api object, etc. Rook can use [aggregated roles](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles) so that only the difference needs to be applied during upgrade instead of completely removing and adding roles with the same names.

**What is use case behind this feature:**
Improved upgrades

**Environment**:

### PR DESCRIPTION
**Pull request by kshlm**
The ClusterRoles in the manifests (and helm chart) for the ceph provider
have been updated to use [aggregated ClusterRoles][1].

All ClusterRoles have been split into 2 ClusterRoles.
An aggregated ClusterRole without any rules, named the same, but with an
aggregationRule that matches labels in the format
'rbac.ceph.rook.io/aggregate-to-<ClusterRole name>: "true"'.
A second ClusterRole which contains the existing rules named
'<ClusterRole name>-rules', with label
'rbac.ceph.rook.io/aggregate-to-<ClusterRole name>: "true"'.

These two ClusterRoles give the same behaviour as the previous
ClusterRole.

Additional rules to a ClusterRole can now be provided by creating a new
ClusterRole with the correct label. This should help make updates
simpler.

[1]: https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles

**Which issue is resolved by this Pull Request:**
Resolves #2634.

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

// known CI issues
[skip ci]

### CODE REVIEW

File: cluster/examples/kubernetes/ceph/csi/rbac/cephfs/csi-provisioner-rbac.yaml [hunk 1]
```diff
@@ -8,6 +8,18 @@ kind: ClusterRole
 apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: cephfs-external-provisioner-runner
+aggregationRule:
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/csi/rbac/cephfs/csi-provisioner-rbac.yaml]: @Madhu-1 Should we update the CSI roles in a separate PR so they can be updated in the ceph-csi repo first? Or shall we go ahead with this change and then port them back to ceph-csi? Either way is fine with me

File: cluster/examples/kubernetes/ceph/upgrade-from-v0.8-create.yaml [hunk 2]
```diff
@@ -214,6 +214,20 @@ metadata:
   labels:
     operator: rook
     storage-backend: ceph
+aggregationRule:
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/upgrade-from-v0.8-create.yaml]: Don't worry about updating the upgrade-from-v0.8 yaml files. @BlaineEXE is working on the 0.9 to 1.0 upgrade guide, which will replace these files.

File: cluster/examples/kubernetes/ceph/upgrade-from-v0.8-replace.yaml [hunk 3]
```diff
@@ -23,6 +23,20 @@ metadata:
   labels:
     operator: rook
     storage-backend: ceph
+aggregationRule:
```
Comment (travisn) [inline cluster/examples/kubernetes/ceph/upgrade-from-v0.8-replace.yaml]: No need to update this file either

**Comment by kshlm** [comment]
> * Helm chart [cluster role template](https://github.com/rook/rook/blob/master/cluster/charts/rook-ceph/templates/clusterrole.yaml)

The Helm charts were updated in my first revision.

> * Integration tests. There is basically another copy of the yaml files in the [test manifests](https://github.com/rook/rook/blob/master/tests/framework/installer/ceph_manifests.go)

I've update these now.

File: cluster/examples/kubernetes/ceph/csi/rbac/cephfs/csi-provisioner-rbac.yaml (hunk 1 above)
Comment (Madhu-1) [reply cluster/examples/kubernetes/ceph/csi/rbac/cephfs/csi-provisioner-rbac.yaml]: @travisn same works for me, @kshlm can you send a PR for the ceph-csi driver with these changes?

**Comment by kshlm** [comment]
I can do that.

**Comment by travisn** [comment]
@kshlm Could you rebase and force push again? This will help trigger the CI again since some issues we were having today.

**Comment by travisn** [comment]
The idea is that in future releases the upgrade will be smoother since we will only need to aggregate the new privs instead of deleting and creating the full set with every release.

**Comment by kshlm** [comment]
I'm not sure why CI failed. Looks like an infra issue. I'll rebase once again to see if CI can complete successfully.

**Comment by travisn** [comment]
@kshlm yes they are known CI issues, thanks

**Pull request merged**